### PR TITLE
feat: ch32v307

### PR DESCRIPTION
**Pull request by Marcuss2**
Add CH32V307 support, note that probe-rs does not have a way to handle the erased flash sequence it has.

Further patch in probe-rs is needed to make it work out of the box. I will be making it soon.

Further improvements would be adding the rest of the CH32V3xx series.

### CODE REVIEW

File: ch32v307/template.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
+name: algorithm-test
+variants:
+  - name: algorithm-test
```
Comment (andelf) [inline ch32v307/template.yaml]:
```suggestion
name: ch32v307
variants:
  - name: CH32V307
```

Although this is not the final algorithm description, better use the chip name for heading.

**Comment by andelf** [comment]
Great job!

The CH32V20x and CH32V30x models have almost identical register blocks. This might also be the case for CH32x035 and CH32L103.

> Further patch in probe-rs is needed to make it work out of the box

I remember attempting this with a CH32V203 before but encountered issues with a frozen MCU. Could you please provide more details?

File: ch32v307/template.yaml (hunk 1 above)
Comment (Marcuss2) [reply ch32v307/template.yaml]: Changed it

**Comment by Marcuss2** [comment]
> Great job!
> 
> The CH32V20x and CH32V30x models have almost identical register blocks. This might also be the case for CH32x035 and CH32L103.
> 
> > Further patch in probe-rs is needed to make it work out of the box
> 
> I remember attempting this with a CH32V203 before but encountered issues with a frozen MCU. Could you please provide more details?

I did encounter possibly an issue like it. The examples provided by WCH and datasheet differ. (The datasheet is the correct one)

Also, probe-rs is very slow at sending commands, likely due to no batching implemented in probe-rs for WCHLink, this leads to algorithm flashing speed of 3.2 KiB/s, survivable for CH32V003, no so much for CH32V307.

Also, the erase check value consists of two bytes 0xe9 0x39, target-gen cannot test it currently, I had to make a dirty hack to test it.

**Comment by andelf** [comment]
I see. Thanks for your explanation.

>  likely due to no batching implemented in probe-rs for WCHLink

CH32V3 family supports memory access and `aampostincrement` in `command` of debug module. But it's not supported by probe-rs. (`unimplemented!("Memory access using abstract commands is not implemted")`)

>  the erase check value consists of two bytes

😢 Yes. It requires probe-rs to change its metadata design.

**Pull request merged**